Fix unchecked pointer Safer CPP warnings in AXTextMarker.cpp

nipunshukla21 opened the pull request:
#### 71cf9beb30395349949eb462ec076c687ee14dd9
<pre>
Fix unchecked pointer Safer CPP warnings in AXTextMarker.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=296090">https://bugs.webkit.org/show_bug.cgi?id=296090</a>
<a href="https://rdar.apple.com/156003843">rdar://156003843</a>

Reviewed by Rupin Mittal and Tyler Wilcock.

Use protectedDocument() over document() functions and make AXTextMarkerRange::toString() use a RefPtr for node in line with Safer CPP guidelines.

* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::TextMarkerData::TextMarkerData):
(WebCore::AXTextMarker::AXTextMarker):
(WebCore::AXTextMarkerRange::AXTextMarkerRange):
(WebCore::AXTextMarkerRange::toString const):

Canonical link: <a href="https://commits.webkit.org/297532@main">https://commits.webkit.org/297532@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/feffcf12f9e8f78fe40ea04c6cbea364d10fc8eb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112054 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31725 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22210 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118077 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62295 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114016 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32401 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40302 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85124 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35789 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115001 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25880 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100823 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65557 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25188 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18963 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61929 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95262 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19038 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121396 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39087 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29085 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93960 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39468 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97073 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93781 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23944 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38995 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16779 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/35109 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38981 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44493 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38618 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41946 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40334 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nipunshukla21 commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -65,7 +65,8 @@ TextMarkerData::TextMarkerData(AXObjectCache& cache, const VisiblePosition& visi
     auto position = visiblePosition.deepEquivalent();
-    auto optionalObjectID = nodeID(cache, position.anchorNode());
+    RefPtr<Node> protectedAnchorNode = position.anchorNode();
+    auto optionalObjectID = nodeID(cache, protectedAnchorNode);
```
I'm unsure as to the whether this is the right move as it seems very contrived but the linter would complain if I tried to use the protectedAnchorNode() function here to return a RefPtr<Node>.

twilco replied:
Huh, yeah surprised this wouldn't work:

```cpp
auto optionalObjectID = nodeID(cache, position.protectedAnchorNode());
```

But I'm assuming that's exactly what you tried and the static analyzer didn't accept it, so not a huge deal.

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -131,8 +132,11 @@ AXTextMarker::AXTextMarker(const CharacterOffset& characterOffset, TextMarkerOri
+    if (!cache)
+        return;
+
+    m_data = cache->textMarkerDataForCharacterOffset(characterOffset, origin);
```
Could do:

```cpp
if (CheckedPtr cache = characterOffset.node->protectedDocument()->axObjectCache())
    m_data = cache->textMarkerDataForCharacterOffset(characterOffset, origin);
```

RupinMittal commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -65,7 +65,8 @@ TextMarkerData::TextMarkerData(AXObjectCache& cache, const VisiblePosition& visi
     auto position = visiblePosition.deepEquivalent();
-    auto optionalObjectID = nodeID(cache, position.anchorNode());
+    RefPtr<Node> protectedAnchorNode = position.anchorNode();
+    auto optionalObjectID = nodeID(cache, protectedAnchorNode);
```
How about this instead:

```
auto optionalObjectID = nodeID(cache, position.protectedAnchorNode().get());
```

RupinMittal commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -48,9 +48,9 @@ WTF_MAKE_TZONE_ALLOCATED_IMPL(AXTextMarkerRange);
 using namespace Accessibility;
 
-static std::optional<AXID> nodeID(AXObjectCache& cache, Node* node)
+static std::optional<AXID> nodeID(AXObjectCache& cache, RefPtr<Node> node)
```
Then we won't need to change this function.

RupinMittal commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -92,7 +93,7 @@ TextMarkerData::TextMarkerData(AXObjectCache& cache, const CharacterOffset& char
 
     treeID = cache.treeID().toUInt64();
-    auto optionalObjectID = nodeID(cache, characterOffsetParam.node.get());
+    auto optionalObjectID = nodeID(cache, characterOffsetParam.node);
```
And then we won't have to do this either.

RupinMittal commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -607,17 +608,18 @@ String AXTextMarkerRange::toString() const
 
         StringBuilder builder;
         for (; !it.atEnd(); it.advance()) {
+            RefPtr<Node> countedNode = it.node();
```
We can do:

```
RefPtr node = it.node();
```

and use that below.

nipunshukla21 replied:
Thanks for the feedback! Out of curiosity, do I only need the template argument when I'm declaring something without assigning it i.e. can I leave it out whenever the variable is being directly/copy initialized?

RupinMittal replied:
No problem! And yes, exactly. 

```
RefPtr node;        // ERROR: need to specify the template argument.
                    // (Because the compiler has no way of deducing the type here)
RefPtr<Node> node;  // OK

RefPtr node = it.node();       // OK
RefPtr<Node> node = it.node()  // OK but the <Node> is not needed. We prefer to be concise.

RupinMittal reviewed (commented):
Hello! Welcome to the team! 

Nice work, here are some suggestions :)

RupinMittal commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -607,17 +607,18 @@ String AXTextMarkerRange::toString() const
 
         StringBuilder builder;
         for (; !it.atEnd(); it.advance()) {
+            RefPtr countedNode = it.node();
```
Nit: WebKit style convention is to simply use the plain-old variable name. No `counted` or anything. In special instances (like when capturing `this` in a lambda), we may use `weakThis` or `protectedThis`. But not for normal variables. You won't see any examples of `RefPtr counted*` in the codebase. We prefer:

`RefPtr node = it.node();`

RupinMittal reviewed (commented):
LGTM with a small nit. (And assuming the bots are happy of course).

nipunshukla21 committed:
Fix unchecked pointer Safer CPP warnings in AXTextMarker.cpp
https://bugs.webkit.org/show_bug.cgi?id=296090
rdar://156003843

Reviewed by Rupin Mittal and Tyler Wilcock.

Use protectedDocument() over document() functions and make AXTextMarkerRange::toString() use a RefPtr for node in line with Safer CPP guidelines.

* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::TextMarkerData::TextMarkerData):
(WebCore::AXTextMarker::AXTextMarker):
(WebCore::AXTextMarkerRange::AXTextMarkerRange):
(WebCore::AXTextMarkerRange::toString const):

Canonical link: https://commits.webkit.org/297532@main
